File: ios/PrivatePensieve/Screens/Talk/TalkViewModel.swift
// TalkViewModel.swift
// Private Pensieve AI — iOS
// State machine for voice flow: idle → recording → transcribing → preview → saved

import SwiftUI
import Combine

/// Talk screen state machine.
/// Controls the full voice flow: idle → listening → transcribing → preview → saved.
@MainActor
final class TalkViewModel: ObservableObject {

    // MARK: - State

    enum TalkState: Equatable {
        case idle
        case recording(duration: TimeInterval)
        case transcribing
        case preview(transcript: String, aiReply: String?)
        case saving
        case saved
        case error(String)
    }

    @Published var state: TalkState = .idle
    @Published var liveTranscript: String = ""
    @Published var recordingDuration: TimeInterval = 0

    // MARK: - Dependencies

    private let aiBrain: AIBrainProvider
    private let stt: SpeechToTextProvider
    private let tts: TTSProvider
    private let memoryCardDAO: MemoryCardDAO
    private let conversationDAO: ConversationDAO

    private var recordingTimer: Timer?

    // MARK: - Init

    init(
        aiBrain: AIBrainProvider = FakeAIBrain(),
        stt: SpeechToTextProvider = FakeSpeechToText(),
        tts: TTSProvider = FakeTTS(),
        memoryCardDAO: MemoryCardDAO = MemoryCardDAO(),
        conversationDAO: ConversationDAO = ConversationDAO()
    ) {
        self.aiBrain = aiBrain
        self.stt = stt
        self.tts = tts
        self.memoryCardDAO = memoryCardDAO
        self.conversationDAO = conversationDAO
    }

    // MARK: - Actions

    /// Start recording voice input.
    func startRecording() {
        guard case .idle = state else { return }
        state = .recording(duration: 0)
        recordingDuration = 0
        liveTranscript = ""

        // Start duration timer
        recordingTimer = Timer.scheduledTimer(withTimeInterval: 0.1, repeats: true) { [weak self] _ in
            Task { @MainActor in
                guard let self = self else { return }
                self.recordingDuration += 0.1
                self.state = .recording(duration: self.recordingDuration)
            }
        }
    }

    /// Stop recording and begin transcription.
    func stopRecording() {
        recordingTimer?.invalidate()
        recordingTimer = nil

        state = .transcribing

        Task {
            do {
                // Simulate transcription (V1: FakeSpeechToText)
                let tempURL = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent("recording.m4a")
                let result = try await stt.transcribe(audioFileURL: tempURL)
                liveTranscript = result.text

                // Generate AI reply
                let reply = try await aiBrain.generateFriendReply(transcript: result.text, recentMemories: [])
                state = .preview(transcript: result.text, aiReply: reply)
            } catch {
                state = .error("Transcription failed: \(error.localizedDescription)")
            }
        }
    }

    /// Save the transcript as a conversation and extract memory cards.
    func saveTranscript(_ editedTranscript: String) {
        state = .saving

        Task {
            do {
                // 1. Save conversation
                let conversationId = UUID().uuidString
                let conversation = Conversation(
                    id: conversationId,
                    createdAt: Date(),
                    sourceType: .voice,
                    userTranscript: editedTranscript,
                    aiReply: nil,
                    audioRetained: false,
                    audioFilePath: nil,
                    isArchived: false,
                    isDeleted: false
                )
                try conversationDAO.insert(conversation)

                // 2. Extract memory cards
                let cards = try await aiBrain.extractMemory(transcript: editedTranscript)
                for card in cards {
                    try memoryCardDAO.insert(card)
                }

                // 3. Generate friend reply
                let reply = try await aiBrain.generateFriendReply(transcript: editedTranscript, recentMemories: cards)

                // 4. Speak the reply
                try await tts.speak(text: reply)

                state = .saved

                // Return to idle after brief delay
                try await Task.sleep(nanoseconds: 2_000_000_000)
                state = .idle
                liveTranscript = ""
            } catch {
                state = .error("Save failed: \(error.localizedDescription)")
            }
        }
    }

    /// Discard the current transcript and return to idle.
    func discard() {
        recordingTimer?.invalidate()
        recordingTimer = nil
        state = .idle
        liveTranscript = ""
        recordingDuration = 0
    }

    /// Format recording duration as MM:SS.
    var formattedDuration: String {
        let minutes = Int(recordingDuration) / 60
        let seconds = Int(recordingDuration) % 60
        return String(format: "%02d:%02d", minutes, seconds)
    }
}

File: android/app/src/main/java/com/privatepensieve/app/screens/TalkViewModel.kt
package com.privatepensieve.app.screens

import androidx.compose.runtime.*
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.privatepensieve.app.models.Conversation
import com.privatepensieve.app.models.MemoryCard
import com.privatepensieve.app.providers.FakeAIBrain
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import java.time.Instant
import java.util.UUID

/**
 * Talk screen ViewModel — state machine for voice flow.
 * idle → recording → transcribing → preview → saved
 */
class TalkViewModel : ViewModel() {

    sealed class TalkState {
        data object Idle : TalkState()
        data class Recording(val durationMs: Long = 0) : TalkState()
        data object Transcribing : TalkState()
        data class Preview(val transcript: String, val aiReply: String?) : TalkState()
        data object Saving : TalkState()
        data object Saved : TalkState()
        data class Error(val message: String) : TalkState()
    }

    var state by mutableStateOf<TalkState>(TalkState.Idle)
        private set

    var liveTranscript by mutableStateOf("")
        private set

    var recordingDurationMs by mutableLongStateOf(0L)
        private set

    private val aiBrain = FakeAIBrain()

    // MARK: - Actions

    fun startRecording() {
        if (state !is TalkState.Idle) return
        state = TalkState.Recording()
        recordingDurationMs = 0
        liveTranscript = ""

        // Simulate recording timer
        viewModelScope.launch {
            while (state is TalkState.Recording) {
                delay(100)
                recordingDurationMs += 100
                state = TalkState.Recording(recordingDurationMs)
            }
        }
    }

    fun stopRecording() {
        state = TalkState.Transcribing

        viewModelScope.launch {
            try {
                // Simulate transcription delay (V1: FakeSTT)
                delay(500)
                val transcript = "I was nervous today about whether this app can work, but I still feel it can help people keep their thoughts private."
                liveTranscript = transcript

                val reply = aiBrain.generateFriendReply(transcript, emptyList())
                state = TalkState.Preview(transcript, reply)
            } catch (e: Exception) {
                state = TalkState.Error("Transcription failed: ${e.message}")
            }
        }
    }

    fun saveTranscript(editedTranscript: String) {
        state = TalkState.Saving

        viewModelScope.launch {
            try {
                // 1. Create conversation
                val conversationId = UUID.randomUUID().toString()

                // 2. Extract memory cards
                val cards = aiBrain.extractMemory(editedTranscript)

                // 3. Generate friend reply
                val reply = aiBrain.generateFriendReply(editedTranscript, cards)

                // TODO: Persist to vault via DAOs when Context is available

                state = TalkState.Saved

                // Return to idle after delay
                delay(2000)
                state = TalkState.Idle
                liveTranscript = ""
            } catch (e: Exception) {
                state = TalkState.Error("Save failed: ${e.message}")
            }
        }
    }

    fun discard() {
        state = TalkState.Idle
        liveTranscript = ""
        recordingDurationMs = 0
    }

    val formattedDuration: String
        get() {
            val seconds = (recordingDurationMs / 1000).toInt()
            val mins = seconds / 60
            val secs = seconds % 60
            return "%02d:%02d".format(mins, secs)
        }
}
